File: C++ Subsystems/04-operator-overloading-and-smart-pointers/Student_Database/StudentDatabase.cpp
#include "StudentDatabase.h"
#include <iostream>
using namespace std;

StudentDatabase::StudentDatabase() {
    count = 0;
}

void StudentDatabase::addStudent(Student s)
{
    if(count >= 50)
        throw DatabaseFullException();

    // Check invalid roll number
    if(s.rollNo <= 0)
        throw InvalidRollNumberException();

    for(int i = 0; i < count; i++)
    {
        if(students[i].rollNo == s.rollNo)
            throw DuplicateRollNoException();
    }

    for(int i = 0; i < 5; i++)
    {
        if(s.marks[i] < 0 || s.marks[i] > 100)
            throw InvalidMarksException();
    }

    students[count++] = s;
}

void StudentDatabase::displayStudents() {

    if(count == 0){
        cout << "Database empty\n";
        return;
    }

    for(int i=0;i<count;i++){
        students[i].display();
        cout << endl;
    }
}

void StudentDatabase::searchStudent(int roll) {

    for(int i=0;i<count;i++){
        if(students[i].rollNo == roll){
            students[i].display();
            return;
        }
    }

    cout << "Student not found\n";
}

File: Java Subsystems/01-abstract-banking-and-exception-pipelines/q4/StudentDatabase.java
public class StudentDatabase {
    private Student[] students;
    private int count;
    private static final int MAX_SIZE = 50;

    public StudentDatabase() {
        students = new Student[MAX_SIZE];
        count = 0;
    }

    public void addStudent(Student s)
        throws DuplicateRollNoException, DatabaseFullException, InvalidRollNoException {
        if (s.getRollNo() < 1 || s.getRollNo() > 50) {
            throw new InvalidRollNoException(s.getRollNo());
        }
        if (count >= MAX_SIZE) {
            throw new DatabaseFullException();
        }
        for (int i = 0; i < count; i++) {
            if (students[i].getRollNo() == s.getRollNo()) {
                throw new DuplicateRollNoException(s.getRollNo());
            }
        }
        students[count++] = s;
        System.out.println("Student added successfully.");
    }

    public void displayAll() {
        if (count == 0) {
            System.out.println("No students in database.");
            return;
        }
        System.out.println("--- All Students ---");
        for (int i = 0; i < count; i++) {
            students[i].display();
        }
    }

    public void searchByRollNo(int rollNo) {
        for (int i = 0; i < count; i++) {
            if (students[i].getRollNo() == rollNo) {
                System.out.println("Student found:");
                students[i].display();
                return;
            }
        }
        System.out.println("No student found with roll number " + rollNo);
    }

    public int getCount() { return count; }
}
